### PR TITLE
AMD MMIO UART fixes

### PR DESCRIPTION
**Pull request by daym**
This fixes the AMD MMIO UART drivers, and adds com2.

First, the THRE check is fixed (the check was inverted).

Then the baud rate is set to 115200.

Then, console output is written to COM2 as well.

### CODE REVIEW

File: src/drivers/uart/src/amdmmio.rs [hunk 1]
```diff
@@ -117,7 +117,7 @@ impl Driver for AMDMMIO {
     fn pwrite(&mut self, data: &[u8], _offset: usize) -> Result<usize> {
         'outer: for (sent_count, &c) in data.iter().enumerate() {
             for _ in 0..RETRY_COUNT {
-                if !self.stat.is_set(STAT::THRE) || !self.stat.is_set(STAT::THRE) {
+                if self.stat.is_set(STAT::THRE) || self.stat.is_set(STAT::THRE) {
```
Comment (orangecms) [inline src/drivers/uart/src/amdmmio.rs:125]: This looks a bit odd to me - as it did before... Is it intended that `THRE` (transmitter holding register empty) is being checked twice, and the formerly defined `TDRE`, which is now `TEMT` (transmitter empty), is not actually used here? Or should the second check actually look at `TEMT`?

Having not too much of an understanding, I am reading the following in a datasheet:

> THRE-flag = "1" indicates that there is no data in the Transmission FIFO buffer register and transmission character is able to be written.
> TEMT-flag becomes "1" when there is no data in the register and Transmission shift register in the
transmission control part.
> Both flags become "0" at writing "0" to transmission FIFO buffer.

File: src/mainboard/amd/romecrb/src/main.rs [hunk 2]
```diff
@@ -214,9 +214,9 @@ pub extern "C" fn _start(fdt_address: usize) -> ! {
     uart0.pwrite(b"Welcome to oreboot\r\n", 0).unwrap();
     debug.init().unwrap();
     debug.pwrite(b"Welcome to oreboot - debug port 80\r\n", 0).unwrap();
```
Comment (orangecms) [inline src/mainboard/amd/romecrb/src/main.rs:216]: To not confuse, when switching to COM2 here, the debug print should probably also say COM2

File: src/drivers/uart/src/amdmmio.rs (hunk 1 above)
Comment (daym) [reply src/drivers/uart/src/amdmmio.rs:125]: >This looks a bit odd to me - as it did before... Is it intended that THRE (transmitter holding register empty) is being checked twice

Ron did that part, and I kept it in there. Maybe there's a weird quirk in the AMD hardware (I haven't seen this quirk before)? Or it's in order to make the loop do more and thus keep the CPU pipeline more full? The latter seems useless since I doubt that the UART is pipelined, too, or as fast as the CPU (so it would matter) in the first place ;)

Anyway, I'll ask him.

>, and the formerly defined TDRE, which is now TEMT (transmitter empty), is not actually used here? 

>Or should the second check actually look at TEMT?

No.

There's a parallel to serial conversion going on in the 16550.

I can only speculate on how, but the standard way to do this is to have an array of D flip flops like that (that is NOT THR--rather it's the shift register which the THR goes into): https://www.globalspec.com/ImageRepository/LearnMore/20166/digital%20parallel%20serial%20convert%20diagram%20fig%2078d6e819311dc46368055b18a318367ac.png

Q0...Q3 are the parallel input bits (in our case those Q_i are connected to THR's outputs). (We currently use 8 bits, though--and actually, it's configurable how many bits in the parallel serial converter we want in the 16550).

The transmission process goes like this:

First step, all the MUXes are set to their bottom input.

Within one rising clock edge [A], all the flop flips will load their respective Q_i into their respective internal storage.

Then, all the MUXes are set to their top input.

For all the other clock edges, for each flip flop, it loads the value from the flip flop to the left of it, respectively, on the rising clock edge. Because of the synchronous design each flip flop will get the data of the previous flip flop with a delay of 1 cycle. That means you only have to wait at the rightmost flip flop's output in order to get the output to put on the serial line.

In any case, standard serial is probably a tiny little more involved than that (adds start and stop bits)--but the idea is the same.

`THRE` signals that the first clock edge [A] is done and the data has been loaded into the parallel serial converter. It does not mean that the parallel serial converter is done actually putting them on the serial line.

`TEMT`signals that the parallel serial converter is done actually putting all the bits on the serial line and is now "empty".

As the data producer, we don't care whether the parallel serial converter is done, we just care whether buffer (THR) space is available or not. if we did check `TEMT`, then that means we'd always wait until the parallel to serial converter is done, which would cause an additional delay of at least 1 cycle (more if the stop bit is done internally by the parallel to serial converter).

So I would not check `TEMT`--it's useless information to us--and also really a hardware implementation detail.

Even if we wanted to make sure that the signal made it to the receiver, then it's also useless, since the receiver probably has a 16550, too, and their receiving serial to parallel converter has a delay, too.

In short, I don't see why `TEMT` is there in the hardware at all. Maybe for debugging it...

File: src/mainboard/amd/romecrb/src/main.rs (hunk 2 above)
Comment (daym) [reply src/mainboard/amd/romecrb/src/main.rs:216]: Oh, I was of the impression that the idea was to make the welcome message say which port the respective output is on (so you can find out what is connected to what on the outside of the chip).

I can change it, no problem.

File: src/drivers/uart/src/amdmmio.rs (hunk 1 above)
Comment (Rwantare) [reply src/drivers/uart/src/amdmmio.rs:125]: You should remove it, if the contents on the left of get evaluated to true, the ones on the right don't get evaluated. If you need to check twice to make sure, then you should loop. The code wasn't right before.

Comment (daym) [reply src/drivers/uart/src/amdmmio.rs:125]: It's not to make sure, it's to keep the pipeline full in the face of a branch misprediction--if it has any use, that is.

Comment (daym) [reply src/drivers/uart/src/amdmmio.rs:125]: I've removed it anyway--if we need it, we can re-add it.

File: src/mainboard/amd/romecrb/src/main.rs (hunk 2 above)
Comment (orangecms) [reply src/mainboard/amd/romecrb/src/main.rs:216]: Oh right, I confused things here - I'm sorry! So we have `uart0` now, `debug` (which is port 0x80), and `p0` (which is now com2). I think the variables names could be adjusted a bit, but that's probably not for now. :) Just like this it's fine now, imo.

**Review by orangecms — DISMISSED** [review]

**Comment by orangecms** [comment]
```
2021-01-30T15:52:31.7045543Z     Finished release [optimized] target(s) in 31.11s
2021-01-30T15:52:31.7421754Z rust-objcopy -O binary -R .bss target/riscv64imac-unknown-none-elf/release/hifive target/riscv64imac-unknown-none-elf/release/bootblob.bin
2021-01-30T15:52:31.9791351Z mkdir -p target/riscv64imac-unknown-none-elf/release
2021-01-30T15:52:31.9808532Z dtc -W no-unit_address_vs_reg fixed-dtfs.dts -O dtb -o target/riscv64imac-unknown-none-elf/release/fixed-dtfs.dtb
2021-01-30T15:52:32.4558100Z TARGET_DIR=target/riscv64imac-unknown-none-elf/release cargo run --target x86_64-unknown-linux-gnu --manifest-path /home/vsts/work/1/s/tools/layoutflash/Cargo.toml -- target/riscv64imac-unknown-none-elf/release/fixed-dtfs.dtb target/riscv64imac-unknown-none-elf/release/image.bin
2021-01-30T15:52:32.5681068Z     Finished dev [unoptimized + debuginfo] target(s) in 0.03s
2021-01-30T15:52:32.5691788Z      Running `/home/vsts/work/1/s/tools/layoutflash/target/x86_64-unknown-linux-gnu/debug/layoutflash target/riscv64imac-unknown-none-elf/release/fixed-dtfs.dtb target/riscv64imac-unknown-none-elf/release/image.bin`
2021-01-30T15:52:34.2516230Z **
2021-01-30T15:52:34.2517711Z ** Output: target/riscv64imac-unknown-none-elf/release/image.bin
2021-01-30T15:52:34.2521593Z **
2021-01-30T15:52:34.2522673Z make[1]: Leaving directory '/home/vsts/work/1/s/src/mainboard/sifive/hifive'
2021-01-30T15:52:34.2523585Z make: Target 'mainboards' not remade because of errors.
2021-01-30T15:52:34.2589476Z ##[error]Bash exited with code '2'.
2021-01-30T15:52:34.2616611Z ##[section]Finishing: Build all mainboards
```

Odd, that looks very random. Seems like we can ignore it.

**Comment by Rwantare** [comment]
```   --> src/main.rs:217:28
    |
217 |     let p0 = &mut AMDMMIO::com2();
    |                            ^^^^
    |                            |
    |                            function or associated item not found in `AMDMMIO`
    |                            help: there is an associated function with a similar name: `com1`

```

Here's the actual failure. For what it's worth I don't have to comment out the MMIO code anymore

**Comment by orangecms** [comment]
Ha, parallel builds... of course, somewhere buried deep in the log:

```
2021-01-30T16:28:59.7250414Z error[E0599]: no function or associated item named `com2` found for struct `AMDMMIO` in the current scope
2021-01-30T16:28:59.7252661Z    --> src/main.rs:217:28
2021-01-30T16:28:59.7253200Z     |
2021-01-30T16:28:59.7254021Z 217 |     let p0 = &mut AMDMMIO::com2();
2021-01-30T16:28:59.7254772Z     |                            ^^^^
2021-01-30T16:28:59.7255247Z     |                            |
2021-01-30T16:28:59.7255769Z     |                            function or associated item not found in `AMDMMIO`
2021-01-30T16:28:59.7256856Z     |                            help: there is an associated function with a similar name: `com1`
2021-01-30T16:28:59.7257624Z 
2021-01-30T16:28:59.7832179Z error: aborting due to previous error
2021-01-30T16:28:59.7832919Z 
2021-01-30T16:28:59.7851407Z For more information about this error, try `rustc --explain E0599`.
2021-01-30T16:28:59.7949462Z error: could not compile `romecrb`
2021-01-30T16:28:59.7950014Z 
2021-01-30T16:28:59.7951051Z Caused by:
2021-01-30T16:28:59.7959374Z   process didn't exit successfully: `rustc --crate-name romecrb --edition=2018 src/main.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C opt-level=z -C panic=abort -C lto -C metadata=a5aee7b4c5872e93 -C extra-filename=-a5aee7b4c5872e93 --out-dir /home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps --target x86_64-unknown-none -L dependency=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps -L dependency=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/release/deps --extern 'noprelude:alloc=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/liballoc-2f078159dde2e2b4.rlib' --extern arch=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libarch-22f1d9f95bdcbfe6.rlib --extern clock=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libclock-f055841a2b766699.rlib --extern 'noprelude:compiler_builtins=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libcompiler_builtins-a4ef50c38148090e.rlib' --extern console=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libconsole-161f3a57da877c4c.rlib --extern 'noprelude:core=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libcore-3a34efb182148546.rlib' --extern heapless=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libheapless-ab8ed1c70a0eba28.rlib --extern model=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libmodel-9f6ed29ab60dfb60.rlib --extern payloads=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libpayloads-df7e307aa67f6088.rlib --extern postcard=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libpostcard-42163ef87d5cb457.rlib --extern print=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libprint-0a278f5e57434ae5.rlib --extern register=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libregister-0440cf911fe375c3.rlib --extern static_ref=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libstatic_ref-7f0897b827842426.rlib --extern uart=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libuart-4cb69c2f4a23b98d.rlib --extern util=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libutil-a8e7d705b630b68f.rlib --extern wrappers=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libwrappers-20d401b87db89467.rlib --extern x86_64=/home/vsts/work/1/s/src/mainboard/amd/romecrb/target/x86_64-unknown-none/release/deps/libx86_64-8a20a30fa82aa157.rlib -Z unstable-options -C link-arg=-Tlink.ld -C link-arg=-nostdlib -C link-arg=-nostartfiles` (exit code: 1)
2021-01-30T16:28:59.7970525Z ../../../../Makefile.inc:53: recipe for target 'target/x86_64-unknown-none/release/romecrb' failed
2021-01-30T16:28:59.7971155Z mkdir -p target/x86_64-unknown-none/release
2021-01-30T16:28:59.7971725Z make[1]: *** [target/x86_64-unknown-none/release/romecrb] Error 101
```

**Comment by daym** [comment]
Whoops, I actually forgot to define amdmmip::com2.

Feel free to add, or i will add it in some hours.

**Comment by orangecms** [comment]
We can't push to your repo, so we'll just relax and wait for it. :) Thank you a lot!

**Comment by daym** [comment]
Added.

**Review by orangecms — APPROVED** [review]

**Review by Rwantare — APPROVED** [review]
